subprojects/language-jvm/src/main/java/org/gradle/api/internal/tasks/compile/daemon/CompilerDaemonClient.java

```diff
@@ -32,7 +35,12 @@ public CompilerDaemonClient(DaemonForkOptions forkOptions, CompilerDaemonWorker
     public <T extends CompileSpec> CompileResult execute(Compiler<T> compiler, T spec) {
         // currently we just allow a single compilation thread at a time (per compiler daemon)
         // one problem to solve when allowing multiple threads is how to deal with memory requirements specified by compile tasks
-        return workerProcess.execute(compiler, spec);
+        BuildOperationWorkerRegistry.Completion workerLease = buildOperationWorkerRegistry.operationStart();
```

Comment:
I'm not so sure about that.
Executing the work should hold a lease, even if it's executed using a previously idle worker daemon.

But yes, we could also see actually forking a new worker daemon as eating up CPU.
In that case `reserveNewClient()` should hold a lease.
WDYT?